### ISSUE TITLE
Make PR checks pass faster

### ISSUE THREAD

**Issue by eugene-manuilov** [google/site-kit-wp#1779]
#### Problem Statement

Currently, PR checks take too much time and sometimes become a bottleneck that wastes a good portion of the execution time intended for a ticket. Knowing that our codebase and the number of unit/e2e tests grow every day, we can anticipate that in the future, it can be a big problem. Even with the current number of code and tests in the repo, full check run can take up to half of an hour or more.

#### Proposed Solution

To solve this problem, I suggest to review which checks we need to run and how. Obviously, the first reaction is "we need to run all checks", but if we take a closer look at it, we will see that there are a number of checks that are not vital for a pull request. It is supplementary checks such as storybooks and zips, unit/e2e tests for nightly/trunk versions of PHP and WordPress. These are all good candidates to be updated to run on demand because they are not always need (I would say they are occausally needed).

In addition to this, we can also move Travis CI checks into GitHub Actions. It can help us make use of github cache/artifacts more efficient accross all checks and save time by eliminating a need to run the same actions multiple times per each check.

#### On-Demand Actions DIY

Unfortunately, there is no way to create GitHub actions that can be run on demand, but I have come up with an idea how to do it using existing tools. The trick is to use PR body to indicate which checks we want to run for this PR. For this, we need to add the following markdown to the bottom of the PR template:

```markdown
## On-demand Checks
- [ ] Build Storybook
- [ ] Build Zip
- [ ] PHP Tests (PHP nightly, WordPress trunk)
- [ ] E2E Tests (WordPress nightly)
```
Next we need to update GitHub Actions to do the following:
1. listen to `pull_request.edit` events
1. add `if` condition to appropriate jobs to check if this check is selected.

Here is an example of such `if` condition for `Build Storybook` actions:

```yaml
jobs:
  storybook:
    runs-on: ubuntu-latest
    if: contains(github.event.pull_request.body, '[x] Build Storybook')
    steps:
      - ...
```
Once we implement it, we will get a nice graphical interface to control which additional checks we want to run for each particular pull request, like this:

![checks](https://user-images.githubusercontent.com/873049/87464253-7c4c1900-c61b-11ea-8f6e-ff2ebcd8dab0.png)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

## QA Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
Thanks for writing this up @eugene-manuilov !

A few initial thoughts I have:
- It's a clever idea to use the PR body with the checklist, that's a neat idea
    - It's probably a bit safer to make the jobs opt-out rather than opt-in with this kind of approach
    - I'm not sure we want the PR author to be able to opt-out of any tests
- Most of the time we will always want a zip built, as this is made available for QA via the Tester plugin
- If we were to conditionally run some jobs per PR, I think it might be easier to bake this into the CI config based on the diff of the PR  
That is, if no JS files or dependencies changed we wouldn't need to lint JS or run JS tests. If no PHP files or dependencies changed, we wouldn't need to run PHPUnit tests for specific versions of WP. The tricky part there are builds for the "latest" version where we always want to test against, even if our code didn't change, because a dependency may have.
- I like the idea of using artifact caching between jobs with GitHub Actions, that's something I've thought about as well

The main pain point I think we have right now is mostly a limited bandwidth to run the tests. Only a few of them take very long (E2E being the longest ~ 12m right now). The problem is that we can only run so many jobs at one time across _all_ builds (~10 I think) and each build currently has 11. This means that even with an empty queue, they all can't start at the same time. Also, jobs that are allowed to fail still have to finish before the build is considered complete on Travis which can cause the build to take a minimum of about 15 minutes.

So if we were able to run more jobs simultaneously this would help quite a bit as right now jobs are often waiting in the queue for a minute or to for another one to finish before it can start. There are some features that GitHub Actions has that Travis does not like Artifacts that could cut down on overall time spent in CI.
E.g. building a zip to use in all E2E jobs once rather than building it in each might save 5 minutes of processing time per-build, but may not make the build time any quicker due to the time it takes and jobs that depend on it need to wait. It would allow for more jobs to run though as CI workers would spend less time doing duplicate effort. I'm guessing job concurrency limits will be less of a factor with GitHub Actions than they are with Travis.

I'll think about this some more but it seems like the biggest gain might be from moving our CI to GitHub Actions and doing some basic optimizations with artifacts.

**Comment by eugene-manuilov** [comment]
Thanks for the feedback, @aaemnnosttv!

> Most of the time we will always want a zip built, as this is made available for QA via the Tester plugin

As zip built is need for QA stage which means that the PR has already passed code review and is successfully merged. Thus we don't need to build zip archive on every push and it can be manually triggered before merging by checking off the appropriate box in the PR body.

> That is, if no JS files or dependencies changed we wouldn't need to lint JS or run JS tests. If no PHP files or dependencies changed, we wouldn't need to run PHPUnit tests for specific versions of WP. 

Yeah, this is a great idea too. I like it.

> The tricky part there are builds for the "latest" version where we always want to test against, even if our code didn't change, because a dependency may have.

Can you give me an example?

> It's probably a bit safer to make the jobs opt-out rather than opt-in with this kind of approach

IMO, it's more convenient to say which checks you want to add to the PR rather than checking off those you don't need.

> I'm not sure we want the PR author to be able to opt-out of any tests

PR authors will be able to opt-out of those 4 tests only (storybook, zip, e2e/php tests for WP development version). All other tests are required and can't/won't be omitted.

> The main pain point I think we have right now is mostly a limited bandwidth to run the tests.

This is what "on-demand checks" can help with. If we don't require running those optional tests on every push, our queue will be shorter and checks can pass faster.

**Comment by eugene-manuilov** [comment]
@aaemnnosttv just want to make sure we are on the same page with checkboxes in the PR body - they mean preferences for the current moment in time. Let's say we find out that we need to build a storybook for my PR to understand what happens, then I will check off appropriate checkbox with my mouse and the storybook action will start. Later if we don't need it anymore, we can turn off the storybook action by removing the tick from that checkbox. In other words it can give us flexibility to decide when we need/want to run optional checks.

**Comment by aaemnnosttv** [comment]
> just want to make sure we are on the same page with checkboxes in the PR body

I get it, it's a good idea; my point was that maybe there's a better way that doesn't require manual toggling for what should run.

> As zip built is need for QA stage which means that the PR has already passed code review and is successfully merged. Thus we don't need to build zip archive on every push and it can be manually triggered before merging by checking off the appropriate box in the PR body.

This zip is actually for testing pre-QA or potentially during code review as it will be deleted as soon as the PR is merged.

> > The tricky part there are builds for the "latest" version where we always want to test against, even if our code didn't change, because a dependency may have.
>
> Can you give me an example?

Sure - We have PHP and E2E test jobs that test against the latest version of WP. Since the latest version can change between builds, we should always run it even if our code didn't change. This only applies to a few jobs so it wouldn't be a big deal to keep them going.

> This is what "on-demand checks" can help with. If we don't require running those optional tests on every push, our queue will be shorter and checks can pass faster.

Yes, although like I said, maybe "conditional checks" would be sufficient. I'm keen to hear thoughts from @felixarntz and @tofumatt as well though.

**Comment by felixarntz** [comment]
@eugene-manuilov @aaemnnosttv Thanks for the detailed ideas and discussion so far! Some thoughts from my end:

* I also believe that the biggest bottleneck here are actually the e2e tests and visual regression tests. We may be able to optimize those, but I believe those will continue to be bottlenecks in itself. So I would love for us to move our entire automated testing infrastructure to move to GitHub actions as I also think this will make those faster - plus it's one less service to rely on.
* I'm not very fond of the idea of letting PR authors granularly control which tests to run. One requirement I'd definitely want to keep as is is that every PR should have to pass _all_ tests before it can be merged. Maybe we can have something like two test modes, like "basic" and "strict"?
    * For example, in "basic" we could do some of the things you suggested, like:
        * Not run e2e and visual regression tests (as those are the biggest bottlenecks).
        * Only run PHPUnit tests if PHP files were changed.
        * Only run Jest tests if JS files were changed.
    * In "strict" mode, everything should run like it does now.
    * For a new PR, it would start running in "basic" mode.
    * And then it would be awesome if we could use the PR body to mark a PR as "this is now ready to be _fully_ tested" (i.e. "strict" mode), and at that point the full testing would run. And the PR should only be able to get merged if the latest commit had the "strict" testing run. I'm not sure we can configure GitHub like that though.
    * Commits against `master` and `develop` should _always_ run "strict" mode.
* I believe the Storybook and ZIP builds should continue running at all times. They're not a major bottleneck right now, and we'll want to keep the latest state of every PR visible in our Storybook on GitHub.io as well as available as ZIP for pre-QA.
* Regarding nightly PHP and WP version tests, could we do something that the Travis-CI (or GH Actions if we migrate) build passes already without having completed those? Since they won't cause it to fail anyway.

**Comment by eugene-manuilov** [comment]
@felixarntz @aaemnnosttv thanks for your comments and thoughts. Felix, the following is how we can implement your basic/strict modes idea. The trick is that we need to use pull request's draft status to determine whether it is basic or strict mode. Here is what we need to do: 

##### Travis CI

1. Migrate Travis CI jobs into GitHub Actions as a single workflow file with multiple jobs. They will be similar to what we have now in Travis.
1. Add a new job (let's call it `prerequisite`) to the workflow file that will do the following:
    * define `outputs` section with two variables: `PHP` and `JS`.
    * define a step that runs the following command to set `PHP` output value:

      ```sh-session
      curl -s ${{ github.event.pull_request.diff_url }} | grep --regexp="^diff.*\.php$" -q && echo "::set-output name=PHP::yes" || echo "::set-output name=PHP::no"
      ```
    * define a step that runs the following command to set `JS` output value:

      ```sh-session
      curl -s ${{ github.event.pull_request.diff_url }} | grep --regexp="^diff.*\.js$" -q && echo "::set-output name=JS::yes" || echo "::set-output name=JS::no"
      ```
1. Update migrated jobs from Travis to depend on `prerequisite` job (using `needs` property).
1. Add `if` conditions to migrated jobs with the following values:
    * **`github.event.pull_request.draft == false || needs.prerequisite.outputs.PHP == 'yes'`** for phpunit tests
    * **`github.event.pull_request.draft == false || needs.prerequisite.outputs.JS == 'yes'`** for js tests
    * **`github.event.pull_request.draft == false`** for VRT, e2e and nightly tests

##### Storybook & ZIP

1. Just update the `if` condition in build jobs to check that `github.event.pull_request.draft` is equal to `false` in order to run the job.

**Comment by aaemnnosttv** [comment]
Thanks for the follow up here @eugene-manuilov - I like the idea of using the draft status to control limit some of the tests before it's ready for review. What do you think @felixarntz ?

Regarding the conditional logic for PHP and JS tests, we also want to make sure that the relevant tests run whenever their respective _dependencies_ or relevant config files change, even if none of the source files changed.

So for PHP we'd want to check if the `composer.lock` file had changed, or if the `package-lock.json` had changed for JS.

As for the diff you're doing, we can do this using `git` alone rather than using `curl` + `grep`.

E.g. for PHP:
```
git diff --name-only HEAD ${{ github.event.pull_request.base.ref }} -- composer.lock google-site-kit.php 'includes/*.php' 'tests/*.php' phpunit.xml.dist
```

and for JS

```
git diff --name-only HEAD ${{ github.event.pull_request.base.ref }} -- package-lock.json 'assets/*.js' 'assets/*.json' 'tests/js/*.js'
```

**Comment by eugene-manuilov** [comment]
> As for the diff you're doing, we can do this using git alone rather than using curl + grep.

@aaemnnosttv `git diff` will require one extra step to checkout repository, otherwise it won't work. `curl` + `grep` just work as is without checkout.

> Regarding the conditional logic for PHP and JS tests, we also want to make sure that the relevant tests run whenever their respective dependencies or relevant config files change, even if none of the source files changed.
> 
> So for PHP we'd want to check if the composer.lock file had changed, or if the package-lock.json had changed for JS.

##### PHP

```sh-session
curl -s ${{ github.event.pull_request.diff_url }} | \
grep --regexp="^diff.*\.php$" \
     --regexp="^diff.*composer\.lock$" \
     --regexp="^diff.*phpunit\.xml\.dist$" \
     -q && \
echo "::set-output name=PHP::yes" || echo "::set-output name=PHP::no"
```

##### JS

```sh-session
curl -s ${{ github.event.pull_request.diff_url }} | \
grep --regexp="^diff.*\.js$" \
     --regexp="^diff.*package-lock\.json$" \
     --regexp="^diff.*assets/.*\.json$" \
     -q && \
echo "::set-output name=JS::yes" || echo "::set-output name=JS::no"
```

**Comment by aaemnnosttv** [comment]
> `git diff` will require one extra step to checkout repository, otherwise it won't work. `curl` + `grep` just work as is without checkout.

Ah you're right, I forgot we'd need to do that. If the matching is just as reliable then that works for me 👍

**Comment by aaemnnosttv** [comment]
I just opened #1865 to move VRT to GHA to make this a bit more of an incremental step (any jobs we can shave off of Travis right now are still beneficial).

One thing that I realized that I wanted to mention here though is that GitHub Actions does not provide a way to re-run a specific job - you can only re-run all jobs in a particular workflow. For that reason, things like E2E stability will be very important to address before those jobs can be migrated to GHA.

**Comment by aaemnnosttv** [comment]
@eugene-manuilov is there anything left to do here? I think most if not all of the things we agreed on have been done in other issues. If there is, we should probably create new issues for each part and close this out, otherwise I think we can close this as done 🙂

**Comment by eugene-manuilov** [comment]
@aaemnnosttv yes, we are good now. Going to close this issue.
